File: backend/app/core/sentry.py
"""
Sentry error tracking configuration for FastAPI backend
"""

import sentry_sdk
from sentry_sdk.integrations.fastapi import FastApiIntegration
from sentry_sdk.integrations.sqlalchemy import SqlAlchemyIntegration
from sentry_sdk.integrations.redis import RedisIntegration
import os
from datetime import datetime


def init_sentry():
    """Initialize Sentry SDK for the FastAPI application"""

    sentry_dsn = os.getenv("SENTRY_DSN", "")

    if not sentry_dsn:
        print("⚠️  Sentry DSN not configured. Error tracking disabled.")
        return

    environment = os.getenv("ENVIRONMENT", "development")
    app_version = os.getenv("APP_VERSION", "unknown")

    def before_send_sentry(event, hint):
        """Filter and customize events before sending to Sentry"""

        # Don't send info-level events
        if event.get("level") == "info":
            return None

        # Filter out health check endpoint errors
        if "request" in event:
            url = event["request"].get("url", "")
            if "/health" in url or "/metrics" in url:
                return None

        # Filter out 404 errors
        if event.get("tags", {}).get("http_status") == "404":
            return None

        # Don't send if marked as ignored
        exception = hint.get("exc_info")
        if exception:
            exc_type = exception[0]
            if issubclass(exc_type, (KeyboardInterrupt, SystemExit)):
                return None

        # Add timestamp
        event["timestamp"] = datetime.utcnow().isoformat()

        return event

    try:
        sentry_sdk.init(
            dsn=sentry_dsn,
            integrations=[
                FastApiIntegration(),
                SqlAlchemyIntegration(),
                RedisIntegration(),
            ],
            traces_sample_rate=float(os.getenv("SENTRY_TRACES_SAMPLE_RATE", "0.1")),
            profiles_sample_rate=float(os.getenv("SENTRY_PROFILES_SAMPLE_RATE", "0.1")),
            environment=environment,
            release=f"slotfeed@{app_version}",
            before_send=before_send_sentry,
            # Performance monitoring
            enable_tracing=True,
            # Attach stack traces
            attach_stacktrace=True,
            # Send personal information (turn off in production)
            send_default_pii=False,
            # Max breadcrumbs
            max_breadcrumbs=100,
            # Exception backtrace limit
            request_bodies="medium",
        )

        print(f"✅ Sentry initialized for environment: {environment}")

    except Exception as e:
        print(f"❌ Failed to initialize Sentry: {e}")


def capture_exception(exception: Exception, **kwargs):
    """Capture an exception in Sentry"""
    sentry_sdk.capture_exception(exception, **kwargs)


def capture_message(message: str, level="info"):
    """Capture a message in Sentry"""
    sentry_sdk.capture_message(message, level=level)


def set_user_context(user_id: str, email: str = None, username: str = None):
    """Set user context for Sentry tracking"""
    sentry_sdk.set_user({
        "id": user_id,
        "email": email,
        "username": username,
    })


def add_breadcrumb(message: str, category: str = "info", level: str = "info", data: dict = None):
    """Add a breadcrumb for request tracking"""
    sentry_sdk.add_breadcrumb(
        message=message,
        category=category,
        level=level,
        data=data or {},
    )


def set_tags(tags: dict):
    """Set tags for event context"""
    for key, value in tags.items():
        sentry_sdk.set_tag(key, value)


def set_extra(extra: dict):
    """Set extra context for events"""
    sentry_sdk.set_context("extra", extra)

File: frontend/src/lib/sentry.ts
/**
 * Sentry error tracking configuration for Next.js frontend
 */

import * as Sentry from '@sentry/react';
import { BrowserTracing } from '@sentry/tracing';

export function initSentry() {
  const dsn = process.env.NEXT_PUBLIC_SENTRY_DSN;
  const environment = process.env.NEXT_PUBLIC_ENVIRONMENT || 'development';
  const appVersion = process.env.NEXT_PUBLIC_APP_VERSION || 'unknown';

  if (!dsn) {
    console.warn('⚠️  Sentry DSN not configured. Error tracking disabled.');
    return;
  }

  try {
    Sentry.init({
      dsn,
      integrations: [
        new BrowserTracing({
          // Set sampling rate for performance monitoring
          tracingOrigins: [
            'localhost',
            /^\//,
            /^http:\/\/localhost/,
            /^https:\/\/[^/]*\.(slotfeed|localhost)/,
          ],
          routingInstrumentation: Sentry.reactRouterV6Instrumentation(
            window.history
          ),
        }),
        new Sentry.Replay({
          // Mask all text content to avoid capturing sensitive info
          maskAllText: true,
          // Mask all media playback
          blockAllMedia: true,
        }),
      ],
      // Performance monitoring sample rate
      tracesSampleRate: parseFloat(
        process.env.NEXT_PUBLIC_SENTRY_TRACES_SAMPLE_RATE || '0.1'
      ),
      // Session replay sample rate
      replaysSessionSampleRate: parseFloat(
        process.env.NEXT_PUBLIC_SENTRY_REPLAY_SESSION_SAMPLE_RATE || '0.1'
      ),
      // Session replay on error sample rate (100% for error sessions)
      replaysOnErrorSampleRate: 1.0,
      environment,
      release: `slotfeed-frontend@${appVersion}`,
      // Don't send personal information
      sendDefaultPii: false,
      // Maximum number of breadcrumbs
      maxBreadcrumbs: 100,
      // Attach stack traces
      attachStacktrace: true,
      // Custom filter for events
      beforeSend(event, hint) {
        // Filter out certain error types
        const error = hint.error;

        // Ignore certain exceptions
        if (error) {
          if (
            error instanceof Error &&
            (error.name === 'ChunkLoadError' ||
              error.message?.includes('Network request failed') ||
              error.message?.includes('Load failed'))
          ) {
            return null;
          }
        }

        // Only send errors from production or staging
        if (
          environment === 'development' &&
          process.env.NODE_ENV === 'development'
        ) {
          return null;
        }

        // Filter out errors from specific paths
        if (event.request?.url?.includes('/health')) {
          return null;
        }

        return event;
      },
      // Custom breadcrumb filter
      beforeBreadcrumb(breadcrumb, hint) {
        // Filter out noisy breadcrumbs
        if (breadcrumb.category === 'console' && breadcrumb.level === 'debug') {
          return null;
        }

        // Filter out certain navigation breadcrumbs
        if (
          breadcrumb.category === 'navigation' &&
          breadcrumb.data?.to?.includes('/admin')
        ) {
          // Still record admin navigation but with lower priority
          return breadcrumb;
        }

        return breadcrumb;
      },
    });

    console.log(`✅ Sentry initialized for frontend (${environment})`);
  } catch (error) {
    console.error('❌ Failed to initialize Sentry:', error);
  }
}

/**
 * Capture an exception in Sentry
 */
export function captureException(
  error: Error,
  context?: Record<string, any>
) {
  if (context) {
    Sentry.setContext('error-context', context);
  }
  Sentry.captureException(error);
}

/**
 * Capture a message in Sentry
 */
export function captureMessage(
  message: string,
  level: Sentry.SeverityLevel = 'info'
) {
  Sentry.captureMessage(message, level);
}

/**
 * Set user context for tracking
 */
export function setUserContext(
  userId: string,
  email?: string,
  username?: string
) {
  Sentry.setUser({
    id: userId,
    email,
    username,
  });
}

/**
 * Clear user context (e.g., on logout)
 */
export function clearUserContext() {
  Sentry.setUser(null);
}

/**
 * Add a breadcrumb for request tracking
 */
export function addBreadcrumb(
  message: string,
  category: string = 'info',
  level: Sentry.SeverityLevel = 'info',
  data?: Record<string, any>
) {
  Sentry.addBreadcrumb({
    message,
    category,
    level,
    data,
  });
}

/**
 * Set tags for event context
 */
export function setTags(tags: Record<string, string>) {
  Object.entries(tags).forEach(([key, value]) => {
    Sentry.setTag(key, value);
  });
}

/**
 * Set extra context for events
 */
export function setExtra(key: string, value: any) {
  Sentry.setContext(key, value);
}

/**
 * Start a new transaction
 */
export function startTransaction(
  name: string,
  op: string = 'http.client'
): Sentry.Transaction {
  return Sentry.startTransaction({
    name,
    op,
  });
}

/**
 * Wrap an async function with error tracking
 */
export async function withErrorTracking<T>(
  fn: () => Promise<T>,
  context?: Record<string, any>
): Promise<T> {
  try {
    return await fn();
  } catch (error) {
    if (error instanceof Error) {
      captureException(error, context);
    }
    throw error;
  }
}

/**
 * Create a fetch wrapper that tracks API calls in Sentry
 */
export async function trackedFetch(
  url: string,
  options?: RequestInit
): Promise<Response> {
  const transaction = startTransaction(`fetch ${new URL(url).pathname}`, 'http.client');

  try {
    const response = await fetch(url, options);

    transaction.setData('status_code', response.status);
    transaction.setData('url', url);
    transaction.setData('method', options?.method || 'GET');

    if (!response.ok) {
      captureMessage(`API Error: ${response.status} ${url}`, 'error');
    }

    return response;
  } catch (error) {
    if (error instanceof Error) {
      captureException(error, {
        url,
        method: options?.method || 'GET',
      });
    }
    throw error;
  } finally {
    transaction.finish();
  }
}

/**
 * Get Sentry hub for advanced operations
 */
export function getSentryHub(): Sentry.Hub {
  return Sentry.getCurrentHub();
}

export default Sentry;
